### ISSUE TITLE
`useStyle` hook doesn't work for some targets

### ISSUE THREAD

**Issue by ohkimur** [BuilderIO/mitosis#1124]
### I am interested in helping provide a fix!

Yes

### Which generators are impacted?

- [ ] All
- [ ] Angular
- [ ] HTML
- [ ] Qwik
- [X] React
- [X] React-Native
- [X] Solid
- [ ] Stencil
- [ ] Svelte
- [ ] Vue
- [ ] Web components

### Reproduction case

https://mitosis.builder.io/?outputTab=E4UwhgxgLkA%3D&code=JYWwDg9gTgLgBAbzgVwM4FMDKMCGN0A0KG2AngDbpwC+cAZlBCHAEQACARssOQCbpQAdMAgB6EMBgRUwVCwDcAKEXoAHpFhx+dHMnLw6yAHYBjGCKNwAsqQDCTSEfRGYACjCMwqAJSJFcOBMII1R4AG0jHBBCOAwYADko9ABdOABeYixcfFcWbHQAN3QWbyV/TLJKVwADcoCOCF5SPwDWwIhyaAAuOAByDnJkdF66mnLq72UAqHQYZChLV1GAHl5gAoA+UYDl4CMwZBht1pNUVDSEBGO2oM6oHpYZ3hYCa5pqN4KcQfQLyOiPm02sFbAALHBGADmvwQrkKzhgvjSG1is0S0ThRRcglwUGhMEEXx+3kBQNEWyBcAAEuhyJ0AIRwACSgQhcCgxjgezgACV0DgzEQAGpDIiYDrAXhEaBwAAywAAjtxePSVqI1ptyqVFNQgA

### Expected Behaviour

The `useStyle` hook should work for all targets. However, that's not the case for `react`, particularly with `styled components`, or `emotion`. Also, the same problem appears for react native, and solid targets.

When using the `useStyle` hook, the additional CSS must be included in the output target. When compiling to `react` with `styled-components` or `emotion` we could use something like global styles or similar.

### Actual Behaviour

The `useStyle` hook is ignored for `react` target when using `styled-components` or emotion. It also doesn't work with `react-native` and `solid` targets.

### Additional Information

The `useStyle` hook is also ignored when the output is Mitosis.

**Comment by ohkimur** [comment]
This issue is related with #758

**Comment by ohkimur** [comment]
@samijaber If you have a good way to solve this, I'd love to make a PR.

**Comment by samijaber** [comment]
@ohkimur  Something is wrong with your playground links (not just this one, all the ones you share). They end up with broken code in the playground.


Here is what I see:
<img width="1728" alt="CleanShot 2023-04-18 at 11 37 08@2x" src="https://user-images.githubusercontent.com/1393142/232811373-7a6df9c4-ca39-4d14-9adf-f416d9f8d57c.png">


Would you mind sharing a working playground link for me to take a look?

**Comment by ohkimur** [comment]
> @ohkimur Something is wrong with your playground links (not just this one, all the ones you share). They end up with broken code in the playground.
> 
> Here is what I see: <img alt="CleanShot 2023-04-18 at 11 37 08@2x" width="1728" src="https://user-images.githubusercontent.com/1393142/232811373-7a6df9c4-ca39-4d14-9adf-f416d9f8d57c.png">
> 
> Would you mind sharing a working playground link for me to take a look?

Yeah, I noticed that too. I have to look into why that's happening.

**Comment by ohkimur** [comment]
> I updated your playground link to a working one.
> 
> The issue is that, for `styled-components` and `emotion`, we compute styles per-child-block, whereas in other cases we have a root-level CSS that includes the `useStyle` contents.
> 
> * `useStyle` gets parsed into `json.style`
> * `collectCss` is used in most cases and appends `json.style` if it exists: https://github.com/BuilderIO/mitosis/blob/999cc29e48666c40adeafd47c77802f0c0e941fc/packages/core/src/helpers/styles/collect-css.ts#L80-L86
> * the scenarios you described do not use `collectCss`
> 
> The solution I have in mind is to inject a `<style>{json.style}</style>` in all those cases.
> 
> This means going here:
> 
> https://github.com/BuilderIO/mitosis/blob/999cc29e48666c40adeafd47c77802f0c0e941fc/packages/core/src/generators/react/generator.ts#L423-L435
> 
> And adding something like:
> 
> ```tsx
> {
>   json.style.length && (options.stylesType === 'emotion' || options.stylesType === 'styled-components' || ...others)
>     ? `<style>${json.style}</style>`
>     : ''
> }
> ```
> 
> You'll also need to update the `openFrag` and `closeFrag` checks: they need to return `true` if you are adding this style tag, otherwise you'll end up with invalid JSX if there is 1 root element.

I'll take a look into this and make a PR.

**Comment by ohkimur** [comment]
@samijaber Can you elaborate more on this part, I want to make sure I get it right:

> You'll also need to update the openFrag and closeFrag checks: they need to return true if you are adding this style tag, otherwise you'll end up with invalid JSX if there is 1 root element.

**Comment by samijaber** [comment]
if the output is 1 element. i.e.

```jsx
<div class="foo">hello</div>
```

then `openFrag`/`closeFrag` will not wrap it in a fragment. But if you are now adding an inlined style tag, then you end up with:


```jsx
<>
  <div class="foo">hello</div>
  <style>
    .foo {
      color: red;
    }
  </style>
</>
```

Or else the react component will have invalid JSX. So you need to update the logic in the generator to add fragments if there is an inlined style tag

**Comment by ohkimur** [comment]
@samijaber I see now. Thanks.
